File: leetcode/6.py
# -*- coding: utf-8 -*-
# @Author: gunjianpan
# @Date:   2020-04-03 20:24:06
# @Last Modified by:   gunjianpan
# @Last Modified time: 2020-04-03 20:25:23

"""
6. ZigZag Conversion Medium
The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

P   A   H   N
A P L S I I G
Y   I   R
And then read line by line: "PAHNAPLSIIGYIR"

Write the code that will take a string and make this conversion given a number of rows:

string convert(string s, int numRows);

## Example 1:
Input: s = "PAYPALISHIRING", numRows = 3
Output: "PAHNAPLSIIGYIR"

## Example 2:
Input: s = "PAYPALISHIRING", numRows = 4
Output: "PINALSIGYAHRPI"
Explanation:

P     I    N
A   L S  I G
Y A   H R
P     I

Accepted 424,795 Submissions 1,211,251
"""


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        N = len(s)
        rows = [""] * min(numRows, N)
        curRow, goingDown = 0, False
        for c in s:
            rows[curRow] += c
            if curRow == 0 or curRow == numRows - 1:
                goingDown = not goingDown
            curRow += 1 if goingDown else -1
        return "".join(rows)

File: leetcode/6.java
/*
 * @Author: gunjianpan
 * @Date:   2020-07-20 23:06:08
 * @Last Modified by:   gunjianpan
 * @Last Modified time: 2020-07-20 23:15:56
 */
class Solution {
  public String convert(String s, int numRows) {
    if (numRows == 1)
      return s;

    int n = s.length();
    StringBuilder[] a = new StringBuilder[numRows];
    int p = numRows * 2 - 2;
    // System.out.print(s);
    // char[][] ch=new char[numRows][];

    for (int i = 0; i < n; i++) {
      // a[i%p].append(charAt(i));
      int j = i % p;
      System.out.print(a[0]);
      if (i % p < numRows) {
        if (a[j] == null) {
          a[j] = new StringBuilder();
        }
        a[j].append(s.charAt(i));
      }

      if (i % p >= numRows) {
        if (a[p - j] == null)
          a[p - j] = new StringBuilder();
        a[p - j].append(s.charAt(i));
      }
    }
    return ("12");

    //  System.out.println(a[i]);

    //  return(a);
  }
}